runtime/sys.c

```diff
@@ -434,27 +434,54 @@ CAMLprim value caml_sys_unsafe_getenv(value var)
   return val;
 }
 
-CAMLprim value caml_sys_getenv(value var)
+/* Returns either a string or Val_none on error (i.e. the tag is used to encode
+   error conditions */
```

Comment:
You could also use the new `caml_result` (see caml/mlvalues.h) which would be slightly more bureaucratic but less of an ad-hoc solution.

Reply:
Obviously not materially, but it seems a bit wasteful tbh? The fact that OCaml values are either blocks or integers is fundamental, not ad hoc. `caml_result` nicely allowed the removal of an abuse of the OCaml value representation... the function here is still always returning a legal OCaml value from the GC's perspecive, it's just not a typeable value (yet!).

Reply:
It's ad-hoc in that you setup your own protocol to encode success or failure, and people reading the code have to learn it just for this function. (It's a simple protocol and you describe it in a comment, so it's no big deal and feel free to keep it if you prefer this.) Returning a `caml_result` with a `not_found` exception in the failure case would make it obvious how success and failure is represented, using operations already in use in the runtime. In particular `caml_sys_getenv(var)` would be just `caml_get_value_or_raise(sys_getenv(var))`.